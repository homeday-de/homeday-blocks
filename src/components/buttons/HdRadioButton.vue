<template>
  <div
    @click="select"
    @keyup.enter.space="select"
    class="radioButton"
    role="button"
    tabindex="0"
  >
    <input
      v-if="name && value"
      type="radio"
      :name="name"
      :value="value"
      class="radioButton__input"
      @change="select"
    />
    <!-- Desktop icon -->
    <div class="radioButton__iconWrapper radioButton__iconWrapper--desktop">
      <div
        :style="setBackgroundImage(desktopIcon)"
        class="radioButton__icon"
      />
      <div
        :style="setBackgroundImage(desktopIconHover)"
        class="radioButton__icon--isHover radioButton__icon"
      />
    </div>
    <!-- Mobile icon -->
    <div class="radioButton__iconWrapper radioButton__iconWrapper--mobile">
      <div
        :style="setBackgroundImage(mobileIcon)"
        class="radioButton__icon"
      />
    </div>
    <label :for="name" class="radioButton__label" v-text="label" />
    <HdIcon
      :src="chevronIcon"
      class="radioButton__chevron"
    />
  </div>
</template>

<script>
import HdIcon from 'homeday-blocks/src/components/HdIcon.vue';
import { chevronIcon } from 'homeday-blocks/src/assets/small-icons';

export default {
  name: 'hdRadioButton',
  components: {
    HdIcon,
  },
  props: {
    label: String,
    value: {
      type: [String, Number],
      required: true,
    },
    name: {
      type: String,
      required: true,
    },
    desktopIcon: {
      type: String,
      required: true,
    },
    desktopIconHover: {
      type: String,
      required: true,
    },
    mobileIcon: {
      type: String,
      required: true,
    },
  },
  data() {
    return {
      chevronIcon,
    };
  },
  methods: {
    setBackgroundImage: path => ({ 'background-image': `url(${path})` }),
    select() {
      this.$emit('select', {
        value: this.value,
        name: this.name,
      });
    },
  },
};
</script>
<!-- Add "scoped" attribute to limit CSS to this component only -->
<style scoped lang="scss">
@import 'homeday-blocks/src/styles/mixins.scss';

$iconWidth: 48px;
.radioButton {
  $rB: &;
  position: relative;
  width: 100%;
  cursor: pointer;
  display: flex;
  justify-content: flex-start;
  align-items: center;
  transition: transform .2s ease-in-out, box-shadow .2s ease-in-out;

  &__chevron {
    width: 32px;
    height: 32px;
    position: absolute;
    top: 0;
    right: 0;
    bottom: 0;
    margin: auto;

    @media(min-width: $break-mobile) {
      display: none;
    }

    &::v-deep path {
      fill: $quaternary-color;
    }
  }
  // Border
  &:not(:last-child) {
    @media(max-width: $break-mobile - 1) {
      &:before {
        $leftPos: $iconWidth + $inline-s;
        content: "";
        position: absolute;
        left: $leftPos;
        right: 0;
        height: 1px;
        width: calc(100% - #{$leftPos});
        bottom: 0;
        background: getShade($quaternary-color, 60);
      }
    }
  }

  @media(min-width: $break-mobile) {
    flex-direction: column;
    border-radius: 4px;
    padding: $inset-m;
    box-shadow: 0 6px 12px 0 rgba(0,0,0,.3);

    &:hover, &:focus {
      transform: translateY(-2px);
      box-shadow: 0 9px 12px 0 rgba(0,0,0,.2);
      // border: 4px solid getShade($secondary-color, 110);
      outline: none;
      // margin: -4px;
      &:before {
        content: "";
        position: absolute;
        left: 0;
        right: 0;
        top: 0;
        bottom: 0;
        border-radius: 4px;
        border: 4px solid getShade($secondary-color, 110);
      }
    }
  }
  &__input {
    display: none;
  }
  &__iconWrapper {
    display: none;
    margin-right: $inline-s;
    &--mobile {
      display: block;

      @media(min-width: $break-mobile) {
        display: none;
      }
    }
    &--desktop {
      @media(min-width: $break-mobile) {
        display: block;
      }
    }
  }
  &__icon {
    width: 48px;
    height: 48px;
    background-size: contain;
    background-position: center;
    background-repeat: no-repeat;

    @media(min-width: $break-mobile) {
      width: 96px;
      height: 96px;
      #{$rB}:hover & {
        display: none;
      }
    }

    &--isHover {
      display: none;

      #{$rB}:hover & {
        display: block;
      }
    }
  }
  &__label {
    font-size: 16px;
    line-height: 19px;
  }
}
</style>
