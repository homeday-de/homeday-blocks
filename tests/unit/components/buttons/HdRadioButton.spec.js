import { mount } from '@vue/test-utils';
import HdRadioButton from '@/components/buttons/HdRadioButton.vue';

const desktopIcon = 'desktop-icon-url';
const desktopIconHover = 'desktop-icon-hover-url';
const mobileIcon = 'mobile-icon-url';

describe('HdRadioButton', () => {
  let wrapper;

  const name = 'lorem';
  const value = 'ipsum';
  const label = 'dolor';
  const ICON_SELECTOR = '.radioButton__icon';
  const ICON_HOVER_SELECTOR = '.radioButton__icon--isHover';

  beforeEach(() => {
    wrapper = mount(HdRadioButton, {
      propsData: {
        name,
        value,
        label,
        desktopIcon,
        desktopIconHover,
        mobileIcon,
      },
    });
  });

  test('the component is rendered', () => {
    expect(wrapper.html()).toMatchSnapshot();
  });

  test('shows the provided icon', () => {
    expect(wrapper.find(ICON_SELECTOR).attributes().style).toEqual(expect.stringContaining(desktopIcon));
    expect(wrapper.find(ICON_HOVER_SELECTOR).attributes().style).toEqual(expect.stringContaining(desktopIconHover));
  });

  test('on change, the proper method is invoked', () => {
    const EVENT_EMITTED = 'select';
    const input = wrapper.find('input');
    expect(input.attributes().name).toBe(name);
    expect(input.attributes().value).toBe(value);
    expect(wrapper.find('label').text()).toBe(label);

    // Check events emitted before and after change
    expect(wrapper.emitted(EVENT_EMITTED)).toBeFalsy();
    input.trigger('change');
    expect(wrapper.emitted(EVENT_EMITTED)[0][0]).toMatchObject({ value, name });
  });
});
