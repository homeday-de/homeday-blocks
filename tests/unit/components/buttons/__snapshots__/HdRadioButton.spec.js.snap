// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`HdRadioButton the component is rendered 1`] = `
<div role="button" tabindex="0" class="radioButton"><input type="radio" name="lorem" class="radioButton__input" value="ipsum">
  <div class="radioButton__iconWrapper radioButton__iconWrapper--desktop">
    <div class="radioButton__icon" style="background-image: url(desktop-icon-url);"></div>
    <div class="radioButton__icon--isHover radioButton__icon" style="background-image: url(desktop-icon-hover-url);"></div>
  </div>
  <div class="radioButton__iconWrapper radioButton__iconWrapper--mobile">
    <div class="radioButton__icon" style="background-image: url(mobile-icon-url);"></div>
  </div> <label for="lorem" class="radioButton__label">dolor</label>
  <!---->
</div>
`;
